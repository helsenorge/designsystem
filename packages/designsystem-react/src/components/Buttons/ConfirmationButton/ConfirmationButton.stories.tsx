import React from 'react';
import {withKnobs, select, text, boolean} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import ConfirmationButton from './ConfirmationButton';
import {Icon} from '../../..';
import {allIcons} from '../../../../.storybook/knobs';

const stories = storiesOf('Buttons/ConfirmationButton', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <ConfirmationButton
    variant={select('Variant', ['primary', 'secondary', 'tertiary'], 'primary')}
    disabled={boolean('Disabled', false)}
    href={text('Href (a tag only)', '')}
    asTag={select('As tag', ['button', 'a'], 'button')}>
    {text('Text', 'ConfirmationButton')}
  </ConfirmationButton>
));

stories.add('With icon', () => (
  <ConfirmationButton
    variant={select('Variant', ['primary', 'secondary', 'tertiary'], 'primary')}
    disabled={boolean('Disabled', false)}
    leftIcon={<Icon>{select('Icon', allIcons, 'eye')}</Icon>}>
    {text('Text', 'ConfirmationButton')}
  </ConfirmationButton>
));

stories.add('All variants', () => (
  <>
    <ConfirmationButton variant="primary" disabled={boolean('Disabled', false)}>
      {text('Text (primary)', 'ConfirmationButton')}
    </ConfirmationButton>
    <ConfirmationButton variant="secondary" disabled={boolean('Disabled', false)}>
      {text('Text (secondary)', 'ConfirmationButton')}
    </ConfirmationButton>
    <ConfirmationButton variant="tertiary" disabled={boolean('Disabled', false)}>
      {text('Text (tertiary)', 'ConfirmationButton')}
    </ConfirmationButton>
  </>
));
