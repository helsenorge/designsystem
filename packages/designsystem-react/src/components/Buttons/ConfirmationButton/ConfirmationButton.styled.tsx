import styled, {css} from 'styled-components';
import {color} from '../../../theme';
import {ConfirmationButtonVariants} from '.';

const primaryStyle = css`
  border: 0;
  background-color: ${color('strangulation400')};
  color: white;
  position: relative;
  transition: all 200ms;
  &:hover,
  :active,
  :focus {
    background-color: ${color('strangulation500')};
    box-shadow: 0 0 0 0.25rem ${color('strangulation500')};
  }
  &:disabled {
    pointer-events: none;
    background-color: ${color('scalpel')};
    color: ${color('scab')};
  }
`;

// TODO: Consider moving all the border hacks to a generic function across all buttons
const secondaryStyle = css`
  background-color: transparent;
  border: 0.125rem solid ${color('strangulation400')};
  color: ${color('strangulation400')};
  position: relative;
  padding: 0 2.375rem;
  transition: background-color 200ms;
  &:before {
    content: ' ';
    position: absolute;
    top: -0.25rem;
    left: -0.25rem;
    right: -0.25rem;
    bottom: -0.25rem;
    border: 0.125rem solid transparent;
  }
  &:after {
    content: ' ';
    position: absolute;
    top: -0.375rem;
    left: -0.375rem;
    right: -0.375rem;
    bottom: -0.375rem;
    border: 0.125rem solid transparent;
  }
  &:hover,
  :active,
  :focus {
    background-color: ${color('strangulation500', {translucent: true})};
    border: 0.125rem solid transparent;
    color: ${color('strangulation500')};
    box-shadow: none;
    &:before {
      border: 0.125rem solid ${color('strangulation500')};
    }
    &:after {
      border: 0.125rem solid ${color('strangulation500')};
      transition: border 200ms;
    }
  }
  &:disabled {
    pointer-events: none;
    background-color: transparent;
    border: 0.125rem solid ${color('waitingRoom')};
    color: ${color('scab')};
  }
`;

const tertiaryStyle = css`
  background-color: transparent;
  border: 0.125rem solid transparent;
  color: ${color('strangulation400')};
  position: relative;
  padding: 0 2.375rem;
  transition: background-color 200ms;
  &:before {
    content: ' ';
    position: absolute;
    top: -0.25rem;
    left: -0.25rem;
    right: -0.25rem;
    bottom: -0.25rem;
    border: 0.125rem solid transparent;
  }
  &:after {
    content: ' ';
    position: absolute;
    top: -0.375rem;
    left: -0.375rem;
    right: -0.375rem;
    bottom: -0.375rem;
    border: 0.125rem solid transparent;
  }
  &:hover,
  :active,
  :focus {
    background-color: ${color('strangulation500', {translucent: true})};
    border: 0.125rem solid transparent;
    color: ${color('strangulation500')};
    box-shadow: none;
    &:before {
      border: 0.125rem solid ${color('strangulation500')};
    }
    &:after {
      border: 0.125rem solid ${color('strangulation500')};
      transition: border 200ms;
    }
  }
  &:disabled {
    pointer-events: none;
    background-color: transparent;
    border: 0.125rem solid ${color('waitingRoom')};
    color: ${color('scab')};
  }
`;

// TODO: Consider wrapping all icons as styled components so we can manipulate them by component name
const iconStyle = css`
  padding-left: 0.375rem;
  padding-right: 1.5rem;
  & > svg {
    padding-right: 0.375rem;
  }
`;

export const StyledConfirmationButton = styled('button')<{
  variant: ConfirmationButtonVariants;
  hasIcon: boolean;
  disabled: boolean;
}>`
  outline: none;
  height: 3.125rem;
  font-size: 1.25rem;
  cursor: pointer;
  padding: 0 2.5rem;
  display: flex;
  align-items: center;
  font-family: inherit;
  text-decoration: none;
  ${props => props.variant === 'primary' && primaryStyle};
  ${props => props.variant === 'secondary' && secondaryStyle};
  ${props => props.variant === 'tertiary' && tertiaryStyle};
  ${props => props.hasIcon && iconStyle};
`;
