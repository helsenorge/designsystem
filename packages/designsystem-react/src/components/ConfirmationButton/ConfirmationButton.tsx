import React from 'react';
import cx from 'classnames';
import {ButtonProps, Button} from '../Button';

interface ConfirmationButtonProps extends ButtonProps {
  isLoading?: boolean;
}

// TODO: Move most of the logic out in a generic Button-component that the others inherit.
const ConfirmationButton = React.forwardRef((props: ConfirmationButtonProps, ref: any) => {
  const {children, variant, isLoading = false, ...rest} = props;
  const classes = cx(
    'is-confirmation',
    {[`is-${variant}`]: variant === 'secondary' || variant === 'tertiary'},
    {['is-loading']: isLoading},
  );
  return (
    <Button className={classes} ref={ref} {...rest}>
      {children}
    </Button>
  );
});

export {ConfirmationButton};
