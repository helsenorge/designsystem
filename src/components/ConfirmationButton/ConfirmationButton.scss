// primary
$confirmation-button-primary-color-font: $bone;
$confirmation-button-primary-color-background: $strangulation-400;
// primary:hover
$confirmation-button-primary-hover-color-background: $strangulation-500;
$confirmation-button-primary-hover-color-border: $strangulation-500;
// primary:disabled
$confirmation-button-primary-disabled-color-background: $ash;
$confirmation-button-primary-disabled-color-font: $waiting-room;

// secondary
$confirmation-button-secondary-color-font: $strangulation-500;
$confirmation-button-secondary-color-background: $bone;
$confirmation-button-secondary-color-border: $strangulation-400;
// secondary:hover
$confirmation-button-secondary-hover-color-border: $strangulation-100;
$confirmation-button-secondary-hover-color-background: $strangulation-100;
$confirmation-button-secondary-hover-color-box-shadow: $strangulation-500;
$confirmation-button-secondary-hover-color-font: $strangulation-500;
// secondary:disabled
$confirmation-button-secondary-disabled-color-border: $waiting-room;
$confirmation-button-secondary-disabled-color-font: $waiting-room;

// tertiary
$confirmation-button-tertiary-color-font: $strangulation-500;
$confirmation-button-tertiary-color-background: transparent;
// tertiary:hover
$confirmation-button-tertiary-hover-color-border: $strangulation-100;
$confirmation-button-tertiary-hover-color-background: $strangulation-100;
$confirmation-button-tertiary-hover-color-box-shadow: $strangulation-500;
$confirmation-button-tertiary-hover-color-font: $strangulation-500;
// tertiary:disable
$confirmation-button-tertiary-disabled-color-border: $waiting-room;
$confirmation-button-tertiary-disabled-color-font: $waiting-room;

.is-confirmation {
  background-color: $confirmation-button-primary-color-background;
  color: $confirmation-button-primary-color-font;
  &:disabled,
  &.is-disabled {
    color: $confirmation-button-primary-disabled-color-font;
    background-color: $confirmation-button-primary-disabled-color-background;
  }
  &:hover,
  &.is-hovered {
    background-color: $confirmation-button-primary-hover-color-background;
    box-shadow: 0 0 0 0.25rem $confirmation-button-primary-hover-color-border;
  }
  &:active,
  &.is-active {
    background-color: $confirmation-button-primary-hover-color-background;
    box-shadow: 0 0 0 0.25rem $confirmation-button-primary-hover-color-border;
  }
  &:focus,
  &.is-focused {
    background-color: $confirmation-button-primary-hover-color-background;
    box-shadow: 0 0 0 0.25rem $confirmation-button-primary-hover-color-border;
  }
  &.is-secondary {
    @include button-variant-borders(
      $confirmation-button-secondary-hover-color-border,
      $confirmation-button-secondary-hover-color-box-shadow
    );
    background-color: $confirmation-button-secondary-color-background;
    border: 0.125rem solid $confirmation-button-secondary-color-border;
    color: $confirmation-button-secondary-color-font;
    &:disabled,
    &.is-disabled {
      color: $confirmation-button-secondary-disabled-color-font;
      border-color: $confirmation-button-secondary-disabled-color-border;
    }
    &:hover,
    &.is-hovered {
      background-color: $confirmation-button-secondary-hover-color-background;
      border: 0.125rem solid $confirmation-button-secondary-hover-color-border;
    }
    &:active,
    &.is-active {
      background-color: $confirmation-button-secondary-hover-color-background;
      border: 0.125rem solid $confirmation-button-secondary-hover-color-border;
    }
    &:focus,
    &.is-focused {
      background-color: $confirmation-button-secondary-hover-color-background;
      border: 0.125rem solid $confirmation-button-secondary-hover-color-border;
    }
  }
  &.is-tertiary {
    @include button-variant-borders(
      $confirmation-button-tertiary-hover-color-border,
      $confirmation-button-tertiary-hover-color-box-shadow
    );
    background-color: $confirmation-button-tertiary-color-background;
    color: $confirmation-button-tertiary-color-font;
    &:disabled,
    &.is-disabled {
      color: $confirmation-button-tertiary-disabled-color-font;
      border: 0.125rem solid $confirmation-button-tertiary-disabled-color-border;
    }
    &:hover,
    &.is-hovered {
      background-color: $confirmation-button-tertiary-hover-color-background;
      border: 0.125rem solid $confirmation-button-tertiary-hover-color-border;
      box-shadow: 0 0 0 0.25rem $confirmation-button-tertiary-hover-color-box-shadow;
    }
    &:active,
    &.is-active {
      background-color: $confirmation-button-tertiary-hover-color-background;
      border: 0.125rem solid $confirmation-button-tertiary-hover-color-border;
      box-shadow: 0 0 0 0.25rem $confirmation-button-tertiary-hover-color-box-shadow;
    }
    &:focus,
    &.is-focused {
      background-color: $confirmation-button-tertiary-hover-color-background;
      border: 0.125rem solid $confirmation-button-tertiary-hover-color-border;
      box-shadow: 0 0 0 0.25rem $confirmation-button-tertiary-hover-color-box-shadow;
    }
  }
}
