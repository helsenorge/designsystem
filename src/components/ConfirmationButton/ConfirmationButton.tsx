import React from 'react';
import cn from 'classnames';

// import './ConfirmationButton.scss';

// TODO: Make this inherit from a set of base variant types in constants.
export type ConfirmationButtonVariant = 'seconary' | 'tertiary' | string | undefined;

interface ConfirmationButtonProps
  extends React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: React.ReactNode;
  variant?: ConfirmationButtonVariant;
  isLoading?: boolean;
}

// TODO: Move most of the logic out in a generic Button-component that the others inherit.
const ConfirmationButton = React.forwardRef((props: ConfirmationButtonProps, ref: any) => {
  const {children, variant, isLoading = false, ...rest} = props;
  const classes = cn(
    'button is-confirmation',
    {[`is-${variant}`]: variant === 'secondary' || variant === 'tertiary'},
    {['is-loading']: isLoading},
  );
  return (
    // TODO: Add a 'as' prop so that the button can be either an 'a'-tag or 'button'-tag.
    <button className={classes} ref={ref} {...rest}>
      <span>{children}</span>
    </button>
  );
});

ConfirmationButton.defaultProps = {
  type: 'button',
};

export {ConfirmationButton};
