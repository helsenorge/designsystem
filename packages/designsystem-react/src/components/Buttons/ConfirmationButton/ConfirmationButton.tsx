import React, {useState} from 'react';
import {StyledConfirmationButton} from './ConfirmationButton.styled';
import {HTMLButtonProps, HTMLAnchorProps} from '../../../constants';
import {useIconColor} from '../../../hooks';
import Button, {ButtonVariants} from '../Button/Button';

export type ConfirmationButtonVariants = 'primary' | 'secondary' | 'tertiary';
export type ConfirmationButtonTag = 'a' | 'button';

interface ConfirmationButtonProps extends HTMLButtonProps, HTMLAnchorProps {
  children: React.ReactNode;
  variant?: ConfirmationButtonVariants;
  leftIcon?: any;
  asTag?: ConfirmationButtonTag;
  isLoading?: boolean;
}

const variantMap = {
  primary: 'fill',
  secondary: 'outline',
  tertiary: 'borderless',
};

const ConfirmationButton = React.forwardRef((props: ConfirmationButtonProps, ref: any) => {
  const {
    children,
    variant = 'primary',
    asTag = 'button',
    isLoading = false,
    leftIcon = null,
    disabled = false,
    ...rest
  } = props;
  const [isHovered, setIsHovered] = useState(false);
  // const iconColor = useIconColor({
  //   variant,
  //   variantColor: 'strangulation400',
  //   defaultColor: 'bone',
  //   onHover: {isToggled: isHovered, color: 'strangulation500'},
  //   onDisable: {isToggled: disabled},
  // });
  return (
    <Button
      // onMouseEnter={() => setIsHovered(true)}
      // onMouseLeave={() => setIsHovered(false)}
      // onFocus={() => setIsHovered(true)}
      // onBlur={() => setIsHovered(false)}
      // hasIcon={leftIcon}
      variant={variantMap[variant] as ButtonVariants}
      size="small"
      color="strangulation"
      disabled={disabled}
      asTag={asTag}
      ref={ref}
      {...rest}>
      {/* {leftIcon ? React.cloneElement(leftIcon, {color: iconColor, size: 38, isHovered}) : null} */}
      {children}
    </Button>
  );
});

export default ConfirmationButton;
