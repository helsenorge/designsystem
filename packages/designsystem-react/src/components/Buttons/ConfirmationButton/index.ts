import ConfirmationButton from './ConfirmationButton';
export default ConfirmationButton;
