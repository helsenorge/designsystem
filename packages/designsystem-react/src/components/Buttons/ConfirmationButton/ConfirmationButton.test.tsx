import React from 'react';
import {render} from '@testing-library/react';
import ConfirmationButton from './ConfirmationButton';

test('renders default variant correctly', (): void => {
  const {container} = render(<ConfirmationButton>ConfirmationButton</ConfirmationButton>);
  expect(container).toMatchSnapshot();
});
