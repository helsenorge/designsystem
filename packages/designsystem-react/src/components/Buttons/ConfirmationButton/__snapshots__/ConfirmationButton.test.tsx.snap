// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default variant correctly 1`] = `
<div>
  <button
    class="sc-AxiKw cXTHF"
    color="pulse"
  >
    <span
      class="sc-AxjAm iMxqKn"
    >
      ConfirmationButton
    </span>
  </button>
</div>
`;
